// esendex/build/test/esendexfake.js
var express = require('express');
var fs_1 = require('fs');
var path_1 = require('path');
var bodyParser = require('body-parser');
function copy(source) {
    var origin = {};
    var keys = Object.keys(source);
    var i = keys.length;
    while (i--) {
        origin[keys[i]] = source[keys[i]];
    }
    return origin;
}
;
var EsendexFake = (function () {
    function EsendexFake() {
        this.app = express();
        this.dispatcherRequests = [];
        var textBodyParser = bodyParser.text();
        this.app.get('/v1.0/accounts', this.AccountsEndpoint.bind(this));
        this.app.get('/v1.0/accounts/01234567-89AB-CDEF-0123-456789ABCDEF', this.AccountEndpoint.bind(this));
        this.app.get('/v1.0/messageheaders', this.MessageHeadersEndpoint.bind(this));
        this.app.get('/v1.0/messageheaders/CDEB3533-1F76-46D7-A2A9-0DAF8290F7FC', this.MessageHeaderEndpoint.bind(this));
        this.app.post('/v1.0/messagedispatcher', textBodyParser, this.MessageDispatcherEndpoint.bind(this));
    }
    EsendexFake.prototype.listen = function (port, cb) {
        this.server = this.app.listen(port, cb);
    };
    EsendexFake.prototype.close = function () {
        this.server.close();
    };
    EsendexFake.prototype.MessageDispatcherEndpoint = function (req, res) {
        this.dispatcherRequests.push({
            headers: copy(req.headers),
            body: req.body
        });
        res.status(200);
        res.set('Content-Type', 'application/xml');
        fs_1.createReadStream(path_1.join(__dirname, '..', '..', 'responses', 'messagedispatcher-post.xml')).pipe(res);
    };
    EsendexFake.prototype.MessageHeaderEndpoint = function (req, res) {
        res.status(200);
        res.set('Content-Type', 'application/xml');
        fs_1.createReadStream(path_1.join(__dirname, '..', '..', 'responses', 'messageheaders-get-single.xml')).pipe(res);
    };
    EsendexFake.prototype.MessageHeadersEndpoint = function (req, res) {
        res.status(200);
        res.set('Content-Type', 'application/xml');
        fs_1.createReadStream(path_1.join(__dirname, '..', '..', 'responses', 'messageheaders-get-all.xml')).pipe(res);
    };
    EsendexFake.prototype.AccountEndpoint = function (req, res) {
        res.status(200);
        res.set('Content-Type', 'application/xml');
        fs_1.createReadStream(path_1.join(__dirname, '..', '..', 'responses', 'accounts-get-single.xml')).pipe(res);
    };
    EsendexFake.prototype.AccountsEndpoint = function (req, res) {
        res.status(200);
        res.set('Content-Type', 'application/xml');
        fs_1.createReadStream(path_1.join(__dirname, '..', '..', 'responses', 'accounts-get-all.xml')).pipe(res);
    };
    return EsendexFake;
})();
exports.EsendexFake = EsendexFake;
